pkg/sql/catalog/nstree/catalog.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package nstree
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/catval"
+	"github.com/cockroachdb/errors"
+)
+
+// Catalog is used to store an in-memory copy of the whole catalog, or a portion
+// thereof.
+type Catalog struct {
+	underlying Map
+}
```

Comment:
This wraps `Map` instead of embedding it because we want to used the underlying by-name and by-id trees and all the lazy initialization machinery, but without using the `Upsert` method on `Map` which doesn't do what we want it to do in this use-case (which is basically replacing `catalog.MapDescGetter` with the additional property of maintaining key ordering)